### PR TITLE
feat: migrate to cloudflare pages and vite (Work in progress)

### PR DESCRIPTION
**Pull request by BhanuNexus**
Migrate to Remix Vite and Cloudflare Pages

Status: Work in progress
Issue: #55

### CODE REVIEW

**Comment by benvinegar** [comment]
Wow, awesome – I took a stab at this for a month+ ago and didn't end up finishing.

Let me know how I can help!

File: app/analytics/collect.test.ts
Comment (BhanuNexus) [inline app/analytics/collect.test.ts:1]: Trying to understand on how to update this tests

**Comment by BhanuNexus** [comment]
Hi @benvinegar, Whenever you get sometime, could you please checkout this branch and test? I don't have data for my cloudflare account, hence not able to test the dashboard completely.

Apart from `collect.test.ts`, remaining looks good in my tests.
FYI - I did npm package updates as well.

**Comment by benvinegar** [comment]
@BhanuNexus I think I addressed the tests.

But I don't know how to actually run this. If I try `npm run dev` it will load the homepage, but `/collect` 404s and trying to load the dashboard URL crashes the server.

It looks like this is converted to Cloudflare Pages (e.g. the `functions` dir), so I also tried `wrangler pages dev` but then I get the following compilation errors:

```
❯ node_modules/wrangler/bin/wrangler.js pages dev

 ⛅️ wrangler 3.61.0
-------------------

✘ [ERROR] 1 error(s) and 0 warning(s) when compiling Worker.



✘ [ERROR] Could not resolve "../build/server"

    [[page]].ts:3:23:
      3 │ import * as build from "../build/server";
        ╵                        ~~~~~~~~~~~~~~~~~


✘ [ERROR] Build failed with 1 error:

  [[page]].ts:3:23: ERROR: Could not resolve "../build/server"
  ```

Maybe I'm missing some steps/instructions?

**Comment by BhanuNexus** [comment]
@benvinegar could you please try below?
```bash
# to generate build/client & build/server
npm run build

# preview
npm run preview
```

**Comment by BhanuNexus** [comment]
> I think I addressed the tests.

Thank you :)

**Comment by benvinegar** [comment]
>@benvinegar could you please try below?

Yep, this gets the server running -- thank you -- but looks like it's producing a bad query because it can't extract the timezone correctly. Poking at this.

**Comment by benvinegar** [comment]
Ok, I think it works: the only remaining issue is figuring out how to inject `VERSION` into the context.

This doesn't seem to ever get a value:

```context.cloudflare.env.CF_PAGES_COMMIT_SHA```

**Comment by benvinegar** [comment]
Here's your PR branch running locally, serving my production dataset:

https://github.com/benvinegar/counterscale/assets/2153/937189e2-c075-42e2-91b9-ca4f88695aef

**Comment by benvinegar** [comment]
Trying to deploy this and I basically just get 500 Internal Error with no visibility into what's going on / no actual HTML returned or anything.

**Comment by BhanuNexus** [comment]
Thank you very much for your changes, sorry for trouble with wrong parameter for timezone.

I assuming there are no secrets set, can you try below?

```bash
# for production
npx wrangler pages secret put CF_ACCOUNT_ID --env production
npx wrangler pages secret put CF_BEARER_TOKEN --env production

# for preview
npx wrangler pages secret put CF_ACCOUNT_ID --env preview
npx wrangler pages secret put CF_BEARER_TOKEN --env preview
```

I'm able to run it here https://counterscaleapp.pages.dev

**Comment by BhanuNexus** [comment]
> This doesn't seem to ever get a value:
> 
> `context.cloudflare.env.CF_PAGES_COMMIT_SHA`
> 
> Any ideas?

This value is available with CF build environment (similar to github actions variables) 
I can see value is available once deployed to pages, sample https://counterscaleapp.pages.dev/

**Comment by benvinegar** [comment]
>I am assuming there are no secrets set, can you try below?

Sorry will try to poke at this tonight. I do wish there was better reporting coming out of pages or a way to capture this without blowing up.

**Comment by benvinegar** [comment]
Yep, that did it 🎉

```
# for preview
npx wrangler pages secret put CF_ACCOUNT_ID --env preview
npx wrangler pages secret put CF_BEARER_TOKEN --env preview
```

I didn't need this I think because it reads `.dev.vars` which contained my local secrets.

---

Okay, so my thinking is: this is basically a major version. It's basically a re-architecture of what I'm used to, and I'm not super comfortable just making this `main`.

So here's my suggestion:

1. I actually tag a `1.0.0` release of Counterscale of `HEAD` on `main`
2. I start a `v2` branch and merge this PR into that branch
3. Going forward, new feature development happens on the `v2` branch
    1. I'll change counterscale.dev to autodeploy from the `v2` branch
4. After some time when I feel it's stable, there are no concerns, etc., I'll merge the v2 branch into `main` and tag a `2.0.0` release

@BhanuNexus Thoughts on this plan?

**Comment by BhanuNexus** [comment]
> Note I didn't need this I think because (I think) it reads `.dev.vars` which already had my local secrets.

yeah, this is an issue and this also makes me think of using Cloudflare Pages settings instead of considering `wrangler.toml`.
If we remove `pages_build_output_dir = "build/client"` from `wrangler.toml` then variables can be set from cloudflare UI, that way it might be easy for people who fork this repo and use.

**Comment by BhanuNexus** [comment]
> Yep, that did it 🎉
> 
> ```
> # for preview
> npx wrangler pages secret put CF_ACCOUNT_ID --env preview
> npx wrangler pages secret put CF_BEARER_TOKEN --env preview
> ```
> 
> ^Note I didn't need this I think because (I think) it reads `.dev.vars` which already had my local secrets.
> 
> Okay, so my thinking is: this is basically a major version. It's basically a re-architecture of what I'm used to, and I'm not super comfortable just making this `main`.
> 
> So here's my suggestion:
> 
> 1. I actually tag a `1.0.0` release of Counterscale of `HEAD` on `main`
> 2. I start a `v2` branch and merge this PR into that branch
> 3. Going forward, new feature development happens on the `v2` branch
>    
>    1. I'll change counterscale.dev to autodeploy from the `v2` branch
> 4. After some time when I feel it's stable, there are no concerns, etc., I'll merge the v2 branch into `main` and tag a `2.0.0` release
> 
> @BhanuNexus Thoughts on this plan?

Totally agreed.
Menwhile, I can update documentation and need to check posibility of moving cloudflare function `/collect` to remix, this will help to reduce dependency on CF and devlopment/debugging will be easy in local.

**Comment by benvinegar** [comment]
Just following up – I've been on vacation, but back now and plan to make these branch changes soon.

> Meanwhile, I can update documentation and need to check posibility of moving cloudflare function /collect to remix, this will help to reduce dependency on CF and devlopment/debugging will be easy in local.

👍

File: app/analytics/collect.ts [hunk 1]
```diff
@@ -130,7 +130,7 @@ interface DataPoint {
 // More here: https://developers.cloudflare.com/analytics/analytics-engine/get-started/#limits
 
 export function writeDataPoint(
-    analyticsEngine: CFAnalyticsEngine,
+    analyticsEngine: AnalyticsEngineDataset,
```
Comment (benvinegar) [inline app/analytics/collect.ts:133]: I think generally speaking, this change isn't really related to migrating to `vite`, and more of a separate personal choice that's been inserted (same with renaming `Environment` to `Env`).

I'll keep it (it's more clear I suppose) but generally it's best to not to couple these changes.

File: package.json [hunk 2]
```diff
@@ -5,67 +5,76 @@
     "type": "module",
     "version": "1.0.0",
     "scripts": {
-        "build": "remix build",
-        "deploy": "wrangler deploy --var VERSION:$(git rev-parse HEAD)",
-        "dev": "remix dev --manual -c \"npm start\"",
+        "dev": "remix vite:dev",
+        "build": "remix vite:build",
+        "prepreview": "remix vite:build",
+        "preview": "wrangler pages dev ./build/client",
```
Comment (benvinegar) [inline package.json:11]: When should I use `npm run preview` and when should I use `npm run dev`? Are the `remix` commands deprecated now?

**Review by benvinegar — APPROVED** [review]

File: app/analytics/collect.ts (hunk 1 above)
Comment (BhanuNexus) [reply app/analytics/collect.ts:133]: Agreed, this `AnalyticsEngineDataset` type is directly from `@cloudflare/workers-types`
So, syntaxes and change tracking are aligned with CF.

Regarding `Env`, I personally prefer `Environment` whatever you have set but cloudflare auto generates types from `wrangler.toml` to `worker-configuration.d.ts` by using `wrangler types`, that is the I have removed `Environment` and reused `Env` from `worker-configuration.d.ts`

File: package.json (hunk 2 above)
Comment (BhanuNexus) [reply package.json:11]: `wrangler pages dev ./build/client` command simulate the CF environment locally like functions
Where as `remix vite:build` just works with remix, usually this command alone will have 404 for `/collect` endpoint of CF functions.

If we manage to migrate `/collect` endpoint to remix, then preview wouldn't be needed anymore. I believe this is doable with minor changes, I will try to get this done in couple of days.

File: app/routes/collect-beta.tsx
Comment (BhanuNexus) [inline app/routes/collect-beta.tsx:1]: Hi @benvinegar, I have previously added this endpoint to test movement of /collect from CF function to remix.

I have tested this with mine https://cs.usevega.com and seems to be working fine, do you want to have this change included with this PR or may be in later iteration?

Comment (benvinegar) [reply app/routes/collect-beta.tsx:1]: @BhanuNexus I started the `v2` branch for this ongoing migration ([here](https://github.com/benvinegar/counterscale/tree/v2)). I'd make a new PR targeting that.

I'm going to close this PR in the meantime.

**Comment by benvinegar** [comment]
Closing because this PR has become the `v2` branch.

**Comment by benvinegar** [comment]
@BhanuNexus All your changes are now on `main` (version is v2.0.0-beta.2).

**Pull request closed without merging**